enh/y indexeddb

max-nextcloud opened the pull request:
- **wip: try y-indexeddb**
- **chore(split) useIndexedDbProvider from Editor.vue**
- **fix(cron): do not reset document**
- **enh(yjs): store baseVersionEtag alongside doc**
- Fixes: #7415


### TODOs

* [x] Think through other file operations.
* [x] Typing without network, closing the editor, opening it again: no content lost
* [x] After closing and opening the editor, undo history is empty
* [x] cleanup yDoc file server-side when file is deleted @mejo- 
* [x] Add `base_version_etag` as custom property to IndexedDB
  * [x] After loading document from IndexedDB, send `baseVersionEtag` as part of create request
  * [x] In case of conflict, use existing manual conflict handling
  * [x] After manual conflict handling, reset y-indexeddb provider for document
  * [x] Fix endless reload loop when base version has been cleared:
    * Open file with version in indexed db but on a different `baseVersionEtag`.
    * Different versions are detected. Reload button is offered.
    * Reload does not change the indexed db content - so conflict persists.
  * [x] migrate change mime type test to playwright
  * [x] debug failing change mime type test.
  * [x] make bc channel depend on base version etag.
  * [x] Test... start editing offline, overwrite file in other tab, see what happens when getting back online @max-nextcloud 
* [x] Handling conflicts when opening file that was changed offline
  * [x] detect conflict
  * [x] handle scenario with no local changes well
  * [x] display conflict when local changes were made
  * [x] adjust conflict dialog wording
  * [x] test conflict resolutions work
* [x] Reconnecting: saved unsaved changes:
  * [x] indicate dirty state
  * [x] autosave when opening dirty file
  * [x] autosave when reconnecting with editing session open
  * [x] push steps that have not been pushed yet
* [x] make "no conflict when uploading identical content" work again (@mejo-)
* [ ] cleanup client side state for deleted files (possibly a follow-up)
* [x] improve wording for conflict resolution buttons (@silverkszlo, possibly a follow-up if too complex)
  * [x] check if the distinction in `readerSource` property for `CollissionResolveDialog` is still needed (see [here](https://github.com/nextcloud/text/blob/d7f2c9bcef14f303846c23d2b1810ab701d877bf/src/components/Editor.vue#L18))
* [x] test in private browser tab and research whether indexeddb is supported there (@silverkszlo)
* [x] keep editor editable when connection issue got detected (@mejo-)

max-nextcloud committed:
wip: try y-indexeddb

For now text is sometimes duplicated

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(split) useIndexedDbProvider from Editor.vue

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(cron): do not reset document

Keep the baseVersionEtag and the editing session around
in case people who are offline connect again later.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
enh(yjs): store baseVersionEtag alongside doc

... and use it to check if the server is still on the same session.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(offline): persist dirty state in indexed db

When reopening a document that was edited offline
it will also be considered dirty now.

Autosave will not kick in yet... As no steps are pushed.
But when closing the file it will be saved.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(test): explore empty changesets

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(rename): use privateMethods for emitError and emitDocumentStateStep

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(cleanup): _getContent alias for serialize

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(refactor): handle open data in websocket polyfill

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): only accept sync protocol and return sync step 2

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
enh(sync): recover automatically from outdated / renamed doc

If no changes have been made offline
clear the indexedDb cache
and reload Editor.vue
to load the latest editing session from the server.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): ensure dirty is updated when saving in onDestroy

Use a setter function that remains functional
even if the component has already been destroyed
and reactivity does not work anymore.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(logging): some optional debug logging

yjs and editor update logging can be very verbose.
Uncomment it but keep it disabled by default.

Can be enabled by setting the corresponding key in OCA.Text - i.e.
`window.OCA.Text.logWebSocketPolyfill = true`
in the browser console.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): actually disable browser broadcast

It was still enabled as we mistyped the `disableBc` option.

We cannot use it right now as the y-websocket room name
does not include the base version etag.

Therefore different tabs may be in different sessions
and would risk syncing never the less.

Observed this when reloading the editor on an outdated session:
* Cleaning up the old yjs document took a while in the beforeDestroy callback.
* The new document was already created.
* The new y-websocket provider issued a sync step 1 via browser broadcast.
* The old y-websocket provider responded and send its full state.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(test): conflict and sync with autoreload

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(type) a few more files

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(test): add initial test for indexed db

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): Cleanup sessions even with unsaved changes

Otherwise the old session would get mixed with the new content
when people come back with old state in indexed-db.
( Run playwright/e2e/indexed-db.spec.ts for an example. )

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(type): EditorFactory and setInitialYjsState

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
enh(sync): updateFromContent()

This could be the basis for automatically merging
changes from an offline editing session
and changes by overwriting the doc.

Signed-off-by: Max <max@nextcloud.com>

mejo- commented:
> * [x] cleanup yDoc file server-side when file is deleted @mejo-

Turns out this already happens and we don't have to change anything:  https://github.com/nextcloud/text/blob/ce8ca645ab64cbf5c0de56fe214bb04d4dfee25b/lib/Listeners/BeforeNodeDeletedListener.php#L42

max-nextcloud committed:
fix(indexedDB): handle conflict with local change

Reconnect to a new session but remember the local change.
Show a conflict between the remote session and the local change.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(log): debug log when attempting to open sync service twice

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(conflict): handle conflict when opening initial session

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(conflict): label buttons properly for unsaved local changes

Conflicts with unsaved changes in indexed db work differently.
When the file is overwritten on the server while editing
the editing session keeps going and the new file version is shown read only.

When opening the file and a conflicting version is found in indexed db
that local version is shown read only allowing to connect to a new session.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(split): reload handling local change in Editor.js

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): make bc channel depend on base version etag

Broadcast is disabled right now.
Make sure two editing sessions for the same file to not mix when enabling it.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(offline): save dirty file when opening

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(offline): autosave with disconnects

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(spell): fix collision typo

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(names): rename requireReconnect to displayConnectionIssue

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
chore(test): save and push offline changes on open

The pushing does not happen yet - thus the test is failing.

Signed-off-by: Max <max@nextcloud.com>

max-nextcloud committed:
fix(sync): push steps from offline changes when opening file

Expose the update mechanism inside the websocket provider.
This is normally triggered by updates of the ydoc.
But as the updates happened while offline in a previous session
we need to trigger it explicitely.

The websocket provider holds the remote state of the document.
So it can compute an update based on this and the current ydoc.

Signed-off-by: Max <max@nextcloud.com>

silverkszlo committed:
fix(sync): wait for indexeddb to load before reading dirty flag (ai code)

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
fix(offline): wait for editor, idb and then save with manualSave=true (ai code)

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
fix(test): push first and then get content of file (ai code)

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
fix(types): add missing mentionSearch parameter to createRichEditor

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
chore(cleanup): remove obsolete imports after rebase

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
chore(lint): npm run prettier and lint

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
fix(test): destroy editor after use

Signed-off-by: silver <s.szmajduch@posteo.de>

silverkszlo committed:
fix(Editor.js): forward slots to inner Editor component (ai code)

Signed-off-by: silver <s.szmajduch@posteo.de>

mejo- commented on src/composables/useIndexedDbProvider.ts:
```diff
@@ -23,14 +23,18 @@ export function useIndexedDbProvider(
+			})
+		})
+		.catch((err) => {
+			logger.warn('IndexedDB sync did not complete', { err })
```
Do we know in which scenarios this throws? I wonder whether this should be handled gracefully or whether this is a scenario that actually should throw as something is broken afterwards.

mejo- committed:
fix(useIndexedDbProvider): don't write on closed connection

There might still be dangling async functions trying to write to
IndexedDB after `clearIndexedDB()` was called. Guard against these race
conditions.

Generated with the help of AI but verified and implemented manually.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
fix(conflict): persist local state in local storage

Needed so local changes don't get lost when editor is closed in
case of conflict.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
fix(documentStatus): display document status on conflict

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
test(playwright): migrate conflict tests from Cypress

Signed-off-by: Jonas <jonas@freesources.org>

silverkszlo committed:
refactor(CollisionResolveDialog): rephrase btns in CollisionResolveDialog

Rename buttons to hopefully make them more understandable.

Signed-off-by: silver <s.szmajduch@posteo.de>

mejo- committed:
fix(document): don't reset document when uploading same content

Fixes 'no conflict when uploading identical content with unsaved
changes'

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
chore(test): move `useDelayedFlag` test to `tests` directory

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
feat(offline): customizable delay before read-only, default 5 minutes

Signed-off-by: Jonas <jonas@freesources.org>

silverkszlo commented on lib/Listeners/NodeWrittenResetDocumentListener.php:
```diff
@@ -0,0 +1,71 @@
+			try {
+				$this->documentService->resetDocument($node->getId(), true);
+			} catch (DocumentHasUnsavedChangesException|NotFoundException $e) {
+				// Do not throw during event handling in this is expected to happen
```
sorry for comment-commenting, but it's phrased in a way that I don't understand what you mean

mejo- replied:
Hehe, this was taken over from the old BeforeNodeWrittenListener. Good question, I think it means: "Do not throw during event handling." and then it lists what happens when. I adjusted the comment accordingly.

mejo- committed:
chore(editor): remove unused property from Wrapper component

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
test(playwright): enable sharding to run tests in parallel in CI

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
test(playwright): longer timeout for first conflict UI check

Requests might take a few seconds before getting triggered once the
conflicting file got uploaded, so let's use a timeout of 10 seconds for
the first conflict resolution UI check in each test.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
test(playwright): improve selector for viewer close button

Prevents failing tests when a toast with close button appears at the
same time.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
fix(WebSocketPolyfill): run onopen callback on initialization

After a long offline period (>70 seconds), y-websocket's internal
watchdog closes the WebSocketPolyfill instance because no messages have
been received.

After going back online, y-websocket creates a new WebSocketPolyfill
instance to reconnect. In the new instance, syncService.open() is
called — but it returns early because the HTTP session is still alive.

This results in `wsconnected` staying false, which means that new steps
never get pushed to the server.

The fix is to run onopen callback explicitly in WebSocketPolyfill's
constructor.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- committed:
fix(PollingBackend): increase poll interval on repeated network errors

When returning to a backgrounded tab, `visibilitychange` resets the fetch
interval to the minimum (300ms). Network errors never called
`increaseRefetchTimer()`, so the aggressive rate persisted indefinitely
while offline, causing ~3 sync requests per second.

Signed-off-by: Jonas <jonas@freesources.org>

mejo- reviewed (approved):
What a ride 🐎

From my side, the PR is ready for merge 🎉

mejo- committed:
chore(idb): use `text-<fileId>` as IDB name

Makes it easier to identify databases created by Text later for cleanup.

Signed-off-by: Jonas <jonas@freesources.org>
